docs: document how to set up Tekton with Pipeline promotion events (#3398)
weaveworks/weave-gitops: website/docs/pipelines/pipelines-with-tekton.mdx

---
title: Pipelines With Tekton
hide_title: true
---

import TierLabel from "./../_components/TierLabel";

import CodeBlock from "@theme/CodeBlock";
import BrowserOnly from "@docusaurus/BrowserOnly";

# Setting Up Pipelines to Trigger a Tekton Pipeline <TierLabel tiers="Enterprise" />

Using Flux's [Notification
Controller](https://fluxcd.io/flux/components/notification/), a [Tekton EventListener](https://tekton.dev/docs/triggers/eventlisteners/) can be triggered on Pipeline promotion events.

## Configuring Tekton Pipeline

### Tekton Tasks

> We have only two tasks to demonstrate how to use parameter values from the

Reviewer: ```suggestion
In this tutorial, we have two tasks to demonstrate how to use parameter values from the
```

> Pipeline event payload, both tasks print out messages with information about the

Reviewer: ```suggestion
Pipeline event payload. Both tasks print out messages with information about the
```

   creates an http listener service.

### Tekton TriggerBinding

A JSON payload from the Notification Service about a Pipeline promotion looks
like this:

```json
{
  "involvedObject": {
    "kind": "Pipeline",
    "namespace": "flux-system",
    "name": "podinfo-pipeline",
    "uid": "74d9e3b6-0269-4c12-9051-3ce8cfb7886f",
    "apiVersion": "pipelines.weave.works/v1alpha1",
    "resourceVersion": "373617"
  },
  "severity": "info",
  "timestamp": "2023-02-08T12:34:13Z",
  "message": "Promote pipeline flux-system/podinfo-pipeline to prod with version 6.1.5",
  "reason": "Promote",
  "reportingController": "pipeline-controller",
  "reportingInstance": "chart-pipeline-controller-8549867565-7822g"
}
```

In out tasks, we are using only the `involvedObject.name`,
`involvedObject.namespace` and `message` fields:

```yaml
---
apiVersion: triggers.tekton.dev/v1beta1
kind: TriggerBinding
metadata:
  name: ww-pipeline-binding
  namespace: ww-pipeline
spec:
  params:
  - name: namespace
    value: $(body.involvedObject.namespace)
  - name: name
    value: $(body.involvedObject.name)
  - name: message
    value: $(body.message)
```

### Tekton TriggerTemplate

> The template has the same parameters as the `Pipeline` resources has:

Reviewer: ```suggestion
The template has the same parameters as the `Pipeline` resources:
```

  namespace: ww-pipeline
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: triggers-example-eventlistener-binding
  namespace: ww-pipeline
subjects:
- kind: ServiceAccount
  name: tekton-ww-pipeline-robot
  namespace: ww-pipeline
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: tekton-triggers-eventlistener-roles
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: triggers-example-eventlistener-clusterbinding
subjects:
- kind: ServiceAccount
  name: tekton-ww-pipeline-robot
  namespace: ww-pipeline
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: tekton-triggers-eventlistener-clusterroles
```

With this `ServiceAccount`, we can create the `EventListener` using the
`TriggerBinding` and `TriggerTemplate`:

```yaml
---
apiVersion: triggers.tekton.dev/v1beta1
kind: EventListener
metadata:
  name: ww-pipeline-listener
  namespace: ww-pipeline
spec:
  serviceAccountName: tekton-ww-pipeline-robot
  triggers:
    - name: ww-pipeline-trigger
      bindings:
      - ref: ww-pipeline-binding
      template:
        ref: ww-pipeline-template
```

> At this point, we should have a `Service` for our `EventListener`.

Author: Not necessarily. Yes, but no. It's enough if they add it to the cluster in any other ways if they have a hybrid solution (even with `kubectl apply`), and to be fair, if they add them all together it still works.

It's most like "see it creates a server and you can inspect it" at any point, for example if they want to `port-forward` it for testing.

I do not expect any guides here to state "and commit it to your giptops repo otherwise it will not be there".

```bash
❯ kubectl get service -n ww-pipeline
NAME                      TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)             AGE
el-ww-pipeline-listener   ClusterIP   10.96.250.23   <none>        8080/TCP,9000/TCP   3d
```

## Configure Notification Provider

In this case, we are using Tekton in the same cluster, so we can use an internal
address to access the `EventListener` service. If they are not in the same

> cluster, exposing the service may required through an ingress or a service mesh.

Reviewer 2: ```suggestion
cluster, exposing the service may be required through an ingress or a service mesh.
```
